client/plots/hierCluster.config.js

```diff
@@ -52,27 +52,26 @@ export async function getPlotConfig(opts = {}, app) {
 	if (!hcTermGroup.lst?.length) {
 		if (!Array.isArray(opts.terms)) throw 'opts.terms[] not array (may show geneset edit ui)'
 
-		const twlst = []
+		const promises = []
 		for (const i of opts.terms) {
-			let tw
-			// FIXME: should not hardcode term type here
-			// hierarchical clustering will need to be performed
-			// on any numeric term.
-			if (typeof i.term == 'object' && i.term.type == 'geneVariant') {
-				// i is already well-formed tw object
-				tw = i
-			} else {
-				// shape i into term{} and nest into tw{}
-				i.type = 'geneVariant'
-				if (!i.gene && !i.name) throw 'no gene or name present'
-				if (!i.gene) i.gene = i.name
-				tw = { term: i }
+			const tw = i.term ? i : { term: i }
+
+			if (!tw.term.type) {
```

Comment:
i feel this is wrong assumption. in what condition will term type be missing?

Reply:
a lot of urls in url.html do not have term type, such as:
http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22pantallMds3%22,%22genome%22:%22hg38%22,%22plots%22:[{%22chartType%22:%22hierCluster%22,%22dataType%22:%22geneExpression%22,%22terms%22:[{%22gene%22:%22APOH%22},{%22gene%22:%22APP%22},{%22gene%22:%22CCND2%22},{%22gene%22:%22COL3A1%22},{%22gene%22:%22COL5A2%22},{%22gene%22:%22CXCL6%22},{%22gene%22:%22FGFR1%22},{%22gene%22:%22FSTL1%22},{%22gene%22:%22ITGAV%22},{%22gene%22:%22JAG1%22},{%22gene%22:%22JAG2%22},{%22gene%22:%22KCNJ8%22},{%22gene%22:%22LPL%22},{%22gene%22:%22LRPAP1%22},{%22gene%22:%22LUM%22},{%22gene%22:%22MSX1%22},{%22gene%22:%22NRP1%22},{%22gene%22:%22OLR1%22},{%22gene%22:%22PDGFA%22},{%22gene%22:%22PF4%22},{%22gene%22:%22PGLYRP1%22},{%22gene%22:%22POSTN%22},{%22gene%22:%22PRG2%22},{%22gene%22:%22PTK2%22},{%22gene%22:%22S100A4%22},{%22gene%22:%22SERPINA5%22},{%22gene%22:%22SLCO2A1%22},{%22gene%22:%22SPP1%22},{%22gene%22:%22STC1%22},{%22gene%22:%22THBD%22},{%22gene%22:%22TIMP1%22},{%22gene%22:%22TNFRSF21%22},{%22gene%22:%22VAV2%22},{%22gene%22:%22VCAN%22},{%22gene%22:%22VEGFA%22},{%22gene%22:%22VTN%22}],%22termgroups%22:[{%22name%22:%22Variables%22,%22lst%22:[{%22id%22:%22subtype%22}]}]}]}

Reply:
then please fix the urls. this is the wrong type of fix that introduces more bugs and troubles

Reply:
thanks the new fix is okay. later should check against clustering group data type (what's the property name?), and detect and prevent mixing up term types in one clustering group

Reply:
updated hiercluster urls in update_hierCluster_url, except 4-gdc, which breaks in master.
now hierCluster term must provide a type